### ISSUE TITLE
avoid overtraining

### ISSUE THREAD

**Issue by wiso** [brendenpetersen/deep-symbolic-optimization#34]
Hello, I have simple 1D function to be fitted, but all the best solution are very overtrained. Example:

![Schermata da 2022-04-28 15-36-36](https://user-images.githubusercontent.com/143389/165793881-272f1212-0ec0-4229-be7f-e51afc8efe49.png)

the red one is proposed by deep-symbolic-optimization, the green one is just a polynominal of 9th degree.

```
{
    "task" : {
        "task_type" : "regression",
        "dataset" : "mydataset.csv",
        "function_set" : ["add", "sub", "mul", "div", "log", "sin", "cos", "exp", "const"]
    },
    "controller": {"max_length": 16},
    "prior": {"length": {"max_": 16}}

}
```

mydataset.csv:
```
9.00e+01, 1.22e-03
9.50e+01, 1.39e-03
1.00e+02, 1.58e-03
1.05e+02, 1.77e-03
1.10e+02, 1.95e-03
1.15e+02, 2.11e-03
1.20e+02, 2.23e-03
1.25e+02, 2.28e-03
1.30e+02, 2.24e-03
1.35e+02, 2.12e-03
1.40e+02, 1.93e-03
1.45e+02, 1.67e-03
1.50e+02, 1.36e-03
1.60e+02, 5.32e-04
1.70e+02, 1.58e-04
1.80e+02, 1.05e-04
1.90e+02, 7.05e-05
2.00e+02, 5.51e-05
2.10e+02, 4.54e-05

```

**Comment by brendenpetersen** [comment]
It would help if you posted the expression, the reward, and other details like how long you trained before discovering the expresion. Hard to believe you could overfit anything with only 16 tokens.

You can also do a simple train-test split to avoid ovrefitting.

**Comment by wiso** [comment]
The expression, ported to numpy is

```                                                 
np.exp(-69.7934 * np.exp(-2.44152784 * np.sin(1.2736765 * x - 1.4101 + 105.01/x)))
```

it took ~30 minutes to find it on a cpu. I haven't saved the reward, but it was > 0.9.

Here another example:

```
(0.05 * np.exp(np.sin(1.30 * x + 1.7)) - 0.020) / (x - 80)
```

**Comment by brendenpetersen** [comment]
Is this real or synthetic data? It's hard to say whether this has truly overfit. Looks like it identified the period which matches the interval between your datapoints, and some function around that to hit amplitudes.

Removing the `const` token would prevent this, as would including datapoints that aren't evenly spaced.

**Comment by wiso** [comment]
These are the values of a physical quantities computed from the theory. The points on the x-axis are equispaced by construction.
By overfit here I mean that the red curve (dso) is fitting very well the points, but it is not able to give a good estimation between the points. I need something much smoother, as the poly9 example.

**Comment by brendenpetersen** [comment]
If they're synthetic, then I would suggest simply adding un-evenly spaced points. That should avoid this overfitting even if you include sin/cos.

Regarding your general question, you only allowed expressions up to length 16. A 9th degree polynomial would take huge amount of tokens. Roughly 9 `add`, 9 `const`, 45 `mul`, and 45 `x1`.

**Comment by wiso** [comment]
Right, I guess you don't have `pow` operator?

**Comment by brendenpetersen** [comment]
You can add `pow` with about 1 line of code...

And yes, my estimate was very much an upper-bound. Just making the point that 16 tokens is not enough to capture a high-degree polynomial.

**Comment by wiso** [comment]
Thanks, any suggestion for my points? They don't seems to have a strange behaviour. For example, is it possible to control the smoothness (avoiding to much changes of the concavity), or to force the function to be monotonic?

**Comment by brendenpetersen** [comment]
Sorry for the delay. I would try things like: making your `X` points unevenly spaced, using a train-validation split (where you train on a portion of your data, and evaluate it on a held-out validation set), getting rid of the `const` token (which can overfit) or getting rid of trig tokens (if you know a priori you want it to be monotonic, then this shouldn't be a design issue).

I don't think there are ways to force it as directly as you want, like changes in concavity or monotonicity. Those would be semantics-based constraints, which are generally infeasible (as opposed to syntactic constraints, which are tractable).

Feel free to re-open if there's something else to discuss, but closing now as I've given the same suggestions a few times.
